dev/tools/test/candidates_test.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:args/command_runner.dart';
+import 'package:dev_tools/candidates.dart';
+import 'package:dev_tools/repository.dart';
+import 'package:file/memory.dart';
+import 'package:platform/platform.dart';
+
+import '../../../packages/flutter_tools/test/src/fake_process_manager.dart';
+import './common.dart';
+
+void main() {
+  group('candidates command', () {
+    const String flutterRoot = '/flutter';
+    const String flutterBinPath = '$flutterRoot/bin/flutter';
+    const String checkoutsParentDirectory = '$flutterRoot/dev/tools/';
+    const String remoteName = 'origin';
+
+    late MemoryFileSystem fileSystem;
+    FakePlatform? platform;
```

Comment:
Why the mix of late and nullable?

Reply:
platform and processManager are nullable because I set them in `createRunner` since they  and then explicitly null it out in the tearDown (a similar test had a bug where test assets were re-used between tests, so I started explicitly nulling out fields after tests that were not guaranteed to be initialized in `setUp()`).

Reply:
I've had a lot of problems with using setup helpers that close over part of the test declarations themselves, it gets very hard to track the active state. I would suggest some modifications:

move `createRunner` to be a top level, so that you need to explicitly pass it all values its using. You could also return some sort of test context object which has fields initialized to what you want them to be. That should let you remove the teardown logic and nullables. 

I do not think you should be defensively programming around test configuration problems by nulling things out

Reply:
that makes sense to me.

Reply:
removed all side effects from `createRunner()`, and instead initialized `processManager` and `platform` directly in the test.